lib/node_modules/@stdlib/plot/table/unicode/lib/render.js

```diff
@@ -0,0 +1,289 @@
+
+
+// MAIN //
+
+/**
+* Renders a table.
+*
+* @private
+* @throws {Error} output must be able to accommodate every column individually
+* @returns {string} rendered table
+*/
+function render() {
+	/* eslint-disable no-invalid-this, no-underscore-dangle */
+	var wrapIndices;
+	var widths;
+	var start;
+	var table;
+	var wraps;
+	var self;
+	var row;
+	var end;
+	var out;
+	var i;
+	var j;
+
+	self = this;
+	out = '';
+
+	if ( !this._columnWidths ) {
+		// Empty table:
+		return out;
+	}
+
+	// Resolve indices to wrap the table at:
+	wrapIndices = resolveWrapping();
+
+	// Render top margin:
+	out += repeat( '\n', this._marginY );
+
+	// Render table:
+	wraps = [];
+	for ( j = 0; j < wrapIndices.length - 1; j++ ) {
+		table = '';
+		start = wrapIndices[ j ];
+		end = wrapIndices[ j + 1 ];
+
+		// Render top border:
+		if ( this._borders !== 'None' ) {
+			widths = slice( this._columnWidths, start, end );
+			table += renderLine( this._borders[ 0 ], widths, this._corners[ 0 ], this._joints[ 1 ], this._corners[ 1 ] ); // eslint-disable-line max-len
+			table += '\n';
+		}
+
+		// Render headers:
+		if ( this._headers ) {
+			row = slice( this._headers, start, end );
+			widths = slice( this._columnWidths, start, end );
+			table += renderRow( row, widths );
+			table += '\n';
+		}
+
+		// Render header separator:
+		if ( this._headers && this._headerSeparator !== 'None' ) {
+			widths = slice( this._columnWidths, start, end );
+			table += renderLine( this._headerSeparator, widths, this._joints[ 4 ], this._joints[ 0 ], this._joints[ 2 ] ); // eslint-disable-line max-len
+			table += '\n';
+		}
+
+		// Render data:
+		if ( this._data ) {
+			for ( i = 0; i < this._data.length; i++ ) {
+				// Render row:
+				row = slice( this._data[ i ], start, end );
+				widths = slice( this._columnWidths, start, end );
+				table += renderRow( row, widths );
+				if ( i !== this._data.length - 1 || this._borders !== 'None' ) {
+					table += '\n';
+				}
+				// Render row separator:
+				if ( this._rowSeparator !== 'None' && i !== this._data.length - 1 ) {
+					widths = slice( this._columnWidths, start, end );
+					table += renderLine( this._rowSeparator, widths, this._joints[ 4 ], this._joints[ 0 ], this._joints[ 2 ] ); // eslint-disable-line max-len
+					table += '\n';
+				}
+			}
+		}
+
+		// Render bottom border:
+		if ( this._borders !== 'None' ) {
+			widths = slice( this._columnWidths, start, end );
+			table += renderLine( this._borders[ 2 ], widths, this._corners[ 3 ], this._joints[ 3 ], this._corners[ 2 ] ); // eslint-disable-line max-len
+		}
+		wraps.push( table );
+	}
+	out += wraps.join( '\n' );
+
+	// Render bottom margin:
+	out += repeat( '\n', this._marginY );
+	return out;
+
+	/**
+	* Resolves columns to wrap the table at (includes bounding indices).
+	*
+	* @private
+	* @throws {Error} output must be able to accommodate every column individually
+	* @returns {Array<number>} list of column indices
+	*/
+	function resolveWrapping() {
```

Comment:
Rather than use a closure, I suggest either figuring out a way to move these to the parent scope or adding them as private methods on the table prototype. Otherwise, each time `render` is invoked, these functions have to be allocated, etc, which will hurt perf.

Reply:
We can move it to the module scope, and just take in the private properties as arguments? Adding them to the table prototype can be avoided (I think?) as these functions are only used when rendering and nowhere else..

Reply:
That is fine, as well.

Reply:
@kgryte Circling back, these methods are heavily dependent on variables defined in the parent function `render`. Moving them to the module scope means that there would be a lot of arguments. Also, these functions like `renderLine` etc would be used throughout the rendering process so they would have to be allocated anyways..? I initially just wrote one mega function with all the steps, I then decided to break them down to make the code cleaner and reduce duplication by making each high level step into a function..